File: skald/services/embedding_service.py
from typing import Literal

from django.conf import settings


class EmbeddingService:
    """
    Service for generating embeddings using various providers.
    Ensures all embeddings are normalized to 2048 dimensions.
    """

    TARGET_DIMENSION = settings.EMBEDDING_VECTOR_DIMENSIONS

    @staticmethod
    def _normalize_embedding(embedding: list[float]) -> list[float]:
        """Pad or validate embedding to match TARGET_DIMENSION"""
        current_dim = len(embedding)

        if current_dim == EmbeddingService.TARGET_DIMENSION:
            return embedding
        elif current_dim < EmbeddingService.TARGET_DIMENSION:
            # Pad with zeros
            return embedding + [0.0] * (EmbeddingService.TARGET_DIMENSION - current_dim)
        else:
            # Vector too large - not supported
            raise ValueError(
                f"Embedding dimension {current_dim} exceeds maximum supported dimension {EmbeddingService.TARGET_DIMENSION}"
            )

    @staticmethod
    def _generate_voyage_embedding(
        content: str, input_type: Literal["document", "query"]
    ) -> list[float]:
        """Generate embedding using Voyage AI"""
        import voyageai

        client = voyageai.Client(api_key=settings.VOYAGE_API_KEY)
        result = client.embed(
            [content],
            model=settings.VOYAGE_EMBEDDING_MODEL,
            input_type=input_type,
            output_dimension=EmbeddingService.TARGET_DIMENSION,
        )
        return result.embeddings[0]

    @staticmethod
    def _generate_openai_embedding(content: str) -> list[float]:
        """Generate embedding using OpenAI"""
        from openai import OpenAI

        client = OpenAI(api_key=settings.OPENAI_API_KEY)

        response = client.embeddings.create(
            input=content,
            model=settings.OPENAI_EMBEDDING_MODEL,
            dimensions=EmbeddingService.TARGET_DIMENSION,
        )
        return response.data[0].embedding

    @staticmethod
    def generate_embedding(
        query: str, usage: Literal["storage", "search"]
    ) -> list[float]:
        """Generate embedding for either storage or search

        Args:
            query: The query to generate an embedding for
            usage: The usage of the embedding, either "storage" or "search". Used to determine the input type for the Voyage AI embedding.

        Returns:
            The embedding as a list of floats
        """
        provider = settings.EMBEDDING_PROVIDER

        if provider == "voyage":
            input_type = "query" if usage == "search" else "document"
            embedding = EmbeddingService._generate_voyage_embedding(
                query, input_type=input_type
            )
        elif provider == "openai":
            embedding = EmbeddingService._generate_openai_embedding(query)
        else:
            raise ValueError(f"Unsupported embedding provider: {provider}")

        return EmbeddingService._normalize_embedding(embedding)

File: memo-processing-server/src/vectorEmbeddings/embeddingService.ts
import {
    EMBEDDING_PROVIDER,
    EMBEDDING_VECTOR_DIMENSION,
    VOYAGE_API_KEY,
    VOYAGE_EMBEDDING_MODEL,
    OPENAI_API_KEY,
    OPENAI_EMBEDDING_MODEL,
    LOCAL_EMBEDDING_MODEL,
} from '../settings'
import { VoyageAIClient } from 'voyageai'
import OpenAI from 'openai'
import { pipeline } from '@huggingface/transformers'

const TARGET_DIMENSION = EMBEDDING_VECTOR_DIMENSION

class EmbeddingService {
    private static normalizeEmbedding(embedding: number[]): number[] {
        const currentDim = embedding.length

        if (currentDim === TARGET_DIMENSION) {
            return embedding
        } else if (currentDim < TARGET_DIMENSION) {
            // Pad with zeros
            return [...embedding, ...Array(TARGET_DIMENSION - currentDim).fill(0)]
        } else {
            throw new Error(`Embedding dimension ${currentDim} exceeds maximum supported dimension ${TARGET_DIMENSION}`)
        }
    }

    private static async generateVoyageEmbedding(content: string, inputType: 'document' | 'query'): Promise<number[]> {
        const client = new VoyageAIClient({ apiKey: VOYAGE_API_KEY })

        const result = await client.embed({
            input: content,
            model: VOYAGE_EMBEDDING_MODEL,
            inputType: inputType,
            outputDimension: TARGET_DIMENSION,
        })

        if (!result.data?.[0] || !result.data?.[0].embedding) {
            throw new Error('Could not generate vector embedding for storage')
        }

        return result.data?.[0].embedding
    }

    private static async generateOpenAIEmbedding(content: string): Promise<number[]> {
        const client = new OpenAI({ apiKey: OPENAI_API_KEY })

        const response = await client.embeddings.create({
            input: content,
            model: OPENAI_EMBEDDING_MODEL,
            dimensions: TARGET_DIMENSION,
        })

        return response.data[0].embedding
    }

    private static async generateTransformerEmbedding(content: string): Promise<number[]> {
        const client = await pipeline('feature-extraction', 'Xenova/' + LOCAL_EMBEDDING_MODEL)

        const response = await client(content, { pooling: 'mean', normalize: true })
        return response.data
    }

    static async generateEmbedding(content: string, usage: 'storage' | 'search'): Promise<number[]> {
        let embedding: number[]

        if (EMBEDDING_PROVIDER === 'voyage') {
            const inputType = usage === 'storage' ? 'document' : 'query'
            embedding = await this.generateVoyageEmbedding(content, inputType)
        } else if (EMBEDDING_PROVIDER === 'openai') {
            embedding = await this.generateOpenAIEmbedding(content)
        } else if (EMBEDDING_PROVIDER === 'local') {
            embedding = await this.generateTransformerEmbedding(content)
        } else {
            throw new Error(`Unsupported embedding provider: ${EMBEDDING_PROVIDER}`)
        }

        return this.normalizeEmbedding(embedding)
    }
}

export { EmbeddingService }
